lgc/patch/PatchCopyShader.cpp

```diff
@@ -599,7 +601,7 @@ Value *PatchCopyShader::loadValueFromGsVsRing(Type *loadTy, unsigned location, u
   Value *loadValue = UndefValue::get(loadTy);
 
   for (unsigned i = 0; i < elemCount; ++i) {
-    Value *ringOffset = calcGsVsRingOffsetForInput(location + i / 4, i % 4, streamId, builder);
+    Value *ringOffset = calcGsVsRingOffsetForInput(location + i / 4, component + i % 4, streamId, builder);
     auto loadElem = builder.CreateIntrinsic(Intrinsic::amdgcn_raw_buffer_load, elemTy,
                                             {
                                                 m_gsVsRingBufDesc, ringOffset,
```

Comment:
Won't the calculation of `location` have to be updated? E.g. if we have `component == 2` and `elemCount == 6` (due to a double vec3).

Reply:
Yes. I think your concerns are right. This is very tricky in GLSL.
For component, only double and dvec2 can specify it. dvec3 and dvec4 are not allowed to specify component qualifier:

_A scalar double will consume two of these components,
and a dvec2 will consume all four components available within a location. A dvec3 or dvec4 can
only be declared without specifying a component._

The spec guarantees that a component qualifier will not make the data type overflow in a location. So double can specify component=0  and component=2 while dvec2 actually can specify component=0. Of course, this is GLSL spec. We are sure to extend this. I can make changes according to your suggestions if you think it is needed.

Reply:
I just dug into the Vulkan spec, and Appendix A (Vulkan Environment for SPIR-V) has this rule:

> The Component decorations must not be used for a 64-bit vector type with more than two components.

I'd suggest to add the spec reference in a comment and add an assertion.

Reply:
Sure. Will do.